docs/user-guide/SPIN.rst

```diff
@@ -0,0 +1,181 @@
+
+.. note::
+   Mcore models use TransformerEngine as a backend, and it tries to find efficient kernels. But depending on the GPU you have it may not find them. If you ever face errors that relate to kernel finding set these variables on top of your script.
+
+   .. code-block:: bash
+
+      export NVTE_MASKED_SOFTMAX_FUSION=0
+      export NVTE_FLASH_ATTN=0
+      export NVTE_FUSED_ATTN=0
+
+Additionally, TransformerEngine is non-deterministic by default, meaning subsequent runs of SPIN using identical parameters will produce different results, which is not ideal for parameter perturbation.
+Helpfully, TransformerEngine exposes a flag to set if you want to guarantee deterministic training runs:
+
+.. code-block:: bash
+
+   export NVTE_ALLOW_NONDETERMINISTIC_ALGO=0
+
+SFT vs Foundational (base) model for SPIN Training
+##################################################
+Unlike DPO and PPO, SPIN was designed to be run on foundational (base) models, that is, models which have only been trained on autoregressive language prediction tasks and not on instruction following tasks.
+However, you can also run SPIN on models which have been SFTed on instruction-based datasets as well, similar to DPO/PPO. Either type of model will work well with SPIN. If you would like to start with a supervised fine tuned model instead of a base model, please see our full guide on how to perform SFT on a Megatron GPT model :ref:`SFT guide <sft>`.
+
+SPIN Model Training
+###################
+
+SPIN training uses the exact same dataset formatting and files as the NeMo-Aligner SFT trainer. Please see the data formatting section of SFT to understand the data format necessary for SPIN :ref:`SFT guide <sft>`
+
+Once your data is processed into the correct format you are ready to begin SPIN training. You must start with a pretrained or SFT trained model. For this section we will use the SFT model trained in the previous step to train the SPIN model.
+For the purposes of the following sections, we'll assuming your training jsonl file is located in ``/path/to/train_spin_format.jsonl`` and your validation jsonl file is located in ``/path/to/valid_spin_format.jsonl``.
+
+For the below parameters, the ``model.spin.ref_policy_kl_penalty`` corresponds to the beta parameter in the SPIN paper, and ``trainer.spin.max_iterations`` corresponds to T (with ``trainer.spin.max_epochs`` per iteration)
+
+.. tab-set::
+
+    .. tab-item:: Terminal
+        :sync: key3
+
+         To run SPIN model training on the terminal directly
+
+         .. code-block:: bash 
+
+            export GPFS="/path/to/nemo-aligner-repo"
+            export TRAIN_DATA_PATH="/path/to/train_spin_format.jsonl"
+            export VALID_DATA_PATH="/path/to/valid_spin_format.jsonl"
+
+            python -u ${GPFS}/examples/nlp/gpt/train_gpt_spin.py \
+               trainer.num_nodes=1 \
+               trainer.devices=8 \
+               ++model.mcore_gpt=true \
```

Comment:
This line can be removed (already in the default config and should never be changed)

Reply:
Fixed, thanks